### PR TITLE
refactor(clickhouse): a client class, and a cluster name on every failure

### PR DESCRIPTION
**Pull request by 0xdeafcafe**
> **Supersedes #6945**, which was the first commit on this branch. That PR is closed in favour of this one rather than merged separately — see *Why these travel together* below. Nothing from it was dropped.

Three changes to how ClickHouse failures are attributed and how the client is assembled. The first makes an error say *which server* refused it; the second turns the package's dead middleware pipeline into a class that states its own ordering; the third deletes the app's hand-rolled resilience wrapper and has the app consume the package instead.

---

## Part 1 — name the cluster on every failure

A ClickHouse error said what failed and never *which server refused it*.

An organization can be routed to its own private instance, so a rejection from a customer's dedicated cluster and one from the shared cluster produced the same log line. During the 2026-08-13 saturation that ambiguity is the whole reason attribution took an afternoon: ~5.6k `TOO_MANY_SIMULTANEOUS_QUERIES` in three hours had to be pinned to a cluster by reading the limit out of the **vendor's error text** — `Maximum: 25` — and matching it against terraform, because 25 is the `private-instance` module default while the shared cluster runs 300.

That is not a diagnosis, it is a coincidence that the two numbers happened to differ.

### Fix

The name already existed and was being thrown away. `CLICKHOUSE_URL__<label>__<orgId>` carries a human label, documented in the code as *"a human-readable customer name (e.g. `acme`), **ignored by code**"*. Keep it, and stamp it on every failure line as `cluster`.

```
ClickHouse query failed   cluster=acme   operation=query  queryId=…
ClickHouse query failed   cluster=shared operation=insert queryId=…
```

**Kept separate from the existing `instance` field rather than replacing it.** `instance` is a *metrics* label; repointing it would move every existing series and break the dashboards built on them. `cluster` only has to be readable by a person looking at one error, so it is additive and defaults to `instance` when a caller has no better name — no field ever becomes absent.

This includes the transient-retry warning. For a transient failure that later succeeds that line is the *only* record emitted, so omitting the cluster there would have left the claim false for exactly the recovered-failure case that dominates the volume.

### Parsing moved to its own module

It is a string split, and it lived in a file that imports Prisma. Testing it meant generating a Prisma client first, which is why its edge cases had no tests at all. Now pinned by `parseRouteKey.unit.test.ts`:

- a label containing the separator still routes on the **trailing** org id — splitting on the first `__` would send `acme__eu` to org `eu`, i.e. route a customer's traffic to a cluster that does not exist
- an unlabelled variable falls back to the org id, so the field is never empty
- a bare prefix or an empty org segment returns null rather than registering an unroutable entry

### Incidental: stops logging a ClickHouse URL verbatim

`getAllClickHouseInstances` logged `{ orgId, url }` on the duplicate-URL branch. A private ClickHouse URL embeds `user:password@host`, so that line put a database password into the log sink — and therefore into Loki. It now logs the cluster name.

Narrow (one branch, fires only on duplicate config) but it is a credential in a log, so it is fixed here rather than filed.

---

## Part 2 — a client class instead of a middleware pipeline

Replaces the package's middleware `compose()` with a `ClickHouseQueryClient` class that holds its collaborators and runs them in a stated order.

The order is unchanged and still load-bearing — it is now written as nesting in one method instead of implied by an array's index:

```
tenant guard   outermost — refuse before spending a span, a slot or a socket
tracer         outside the limiter, so queue time falls inside the span
limiter        OUTSIDE retry, so one slot is held across every attempt
retry          innermost, wrapping the driver
```

Each factory became the class that owns its state or config: `createConcurrencyLimiter` → `ConcurrencyLimiter` (the only mutable state in the package, now named, and readable through `stats()` without running a statement), `tenantGuard` → `TenantGuard.assert()`, `retry` → `RetryPolicy`, `trace` → `QueryTracer`. The pure logic they delegate to — `checkTenantScope`, `runWithRetry`, `describeQueryError` — is untouched, so the transient classifier this shares with the job queue cannot drift.

`pipeline.ts` became `query.ts`, holding the vocabulary every layer shares plus a new `QueryDriver` port. Naming a module for a mechanism it no longer contains is the misleading kind of comment.

### A real bug the conversion surfaced

`RetryPolicy.run` now falls back to the request's own signal when no explicit one is passed. The middleware read `request.signal` directly, so without the fallback a caller passing a request would **silently retry work nobody is waiting for** — the exact failure the option exists to prevent. Two abort tests caught it.

### Why this was worth doing

The pipeline had **zero consumers**. The app imported only `createConcurrencyLimiter`, `poolSizingFromEnv`, `resolvePoolSize`, `runWithRetry` and the logging constants, and hand-rolls retry-reporting, tracing and outcome logging in `app-layer/clients/clickhouse/resilient-client.ts`. So the package was half-used and its central abstraction was dead.

---

## Part 3 — the app consumes it

Part 2 ended with the package sound and its one real consumer still hand-rolling. This closes that: `resilient-client.ts` is deleted, and the vendor-shaped resilience layer — read retry, outcome logging and metrics, the in-band exception guard, cold-scan warning — is now `VendorClientResilience` in the package: a class holding the policy once and applying it to any vendor client via `wrap()`.

The earlier revision of *Not in this PR* said bridging onto `QueryDriver` changes streaming behaviour and needs a design decision. It still does, and this deliberately is not that bridge. The class wraps the vendor's own `query`/`insert` *structurally* — anything with those two methods — so the lazy `ResultSet`, the guard's `.json()` interception, and the vendor interface its 250+ importers hold all carry over unchanged, and the package still never imports `@clickhouse/client`.

What stays app-side is exactly the domain knowledge, injected as ports from `managedClient.ts`: the Prometheus metrics functions, the two logger names, `translateClickHouseQueryError` (HandledErrors and their remediation copy), the cold-scan detector (its table list is pinned to the migrations by a coverage test), and the transient-message list still owned by the event-sourcing classifier so the two layers cannot drift. `createResilientClickHouseClient` survives as that assembly, on the module whose doc already claimed to assemble every policy in one place.

Behaviour is pinned rather than asserted: the three resilient-client test suites moved with their `@scenario` bindings intact and pass unchanged against the new assembly, and a string-literal diff of the old module against the new class shows the only differences are import paths and the logger names, which moved into the assembly.

---

## Part 4 — reporting cannot change what it reports

From review. Three findings turned out to be the same bug in three places: a metric, a log sink and an audit hook are all host code, and each of them runs inside a `catch`. An exception from any one propagated *in place of* the ClickHouse error, so the caller was handed a telemetry failure and never learned what actually broke.

- **Metrics ports.** `incrementCount` throwing on the failure path replaced the ClickHouse error, and `translateQueryError` never ran. Both statement paths were affected.
- **`TenantGuard.onUnscoped`.** The worst-behaved of the three, because of which branch it is on: it runs only where the guard has already decided to *allow*, so a broken audit sink turned every declared-unscoped statement into a refusal — a reporting hook deciding policy.
- **The logging fallback.** A host passing one logger for both ports has `noticeLogger === outcomeLogger`, so the sink asked to report the failure is the sink that just failed, and its second throw escaped too.

The package already stated the rule, in `retry.ts`: *observability must not change what it observes*. What it did not have was a way to hold it. Wrapping each call site is a discipline and the next person to add a metric forgets it, so the ports are wrapped **once, on the way in** — being unable to throw is now a property of the port rather than of the code using it. The outcome sink is the deliberate exception: its failure is worth a line, and that line goes to the guarded notice sink.

### Two more modules, and why

`vendorClient.ts` was over the repo's 300-source-line limit, and the guards above only made it longer. Extracting the pure helpers was not enough on its own (346 lines), so it split on two seams:

- **`statementShape.ts`** — reading facts back out of the vendor's untyped params and rows. Total against the wrong shape, pure, testable without a driver.
- **`statementReporting.ts`** — `StatementReporter`: which sink, which level, which fields, which counter, and the guard above. This is the split worth having. One module decides *what happens* to a statement; the other decides *how it is described*. `vendorClient.ts` is 183 source lines and now reads as policy.

### Named parameters on the new ports

Per the repo rule, and worth doing now rather than later because these are exported: `StatementMetrics`, `translateQueryError`, `ConcurrencyLimiter.run` and `QueryTracer.trace` all take an input object. The platform functions behind those ports are older than this PR and have callers of their own, so `managedClient.ts` adapts onto them at the assembly seam rather than converting them repo-wide.

### Every new test was mutation-checked

Reverting each guard fails exactly the test that claims it (7 tests), and a retry-exhaustion test now pins the `maxRetries + 1` conversion — the previous fail-then-succeed test passed with that off by one in either direction.

---

## Why these travel together

Part 2 was cut from Part 1's branch and edits the same four files. More to the point, a review of the combined branch found a defect that **only exists once both are present**: Part 1 changed the private-route map to hold route *objects*, and `getPrivateClickHouseUrls()` kept its `ReadonlyMap<string, string>` signature. `tasks/clickhouseMigrate.ts` iterates that getter and hands each value straight to goose as a connection string, so on any deployment with a private instance **the migration task would have run against `"[object Object]"`**.

That is a Part 1 bug. Shipping the two PRs separately would have merged it, then fixed it in the follow-up. The getter now projects `.url`, so its contract holds and the routes stay internal to the module.

Two further defects from the same review, both invisible to the tests because nothing on the branch had been typechecked: `tracing.ts` still imported the deleted `QueryMiddleware` (TS2305), and the route map's stored object was missing a required field (TS2345). Also fixed: the client never forwarded the request to `RetryPolicy`, so an `onRetry` callback reading tenant or table got `undefined` and threw into `runWithRetry`'s guard — which swallows it, meaning the retry telemetry would have disappeared silently rather than failing loudly.

## Tests

`client.test.ts` pins what no collaborator's own tests can show, because the nesting is only visible when they run together:

- a refused statement never reaches the driver
- a retrying statement **keeps** its slot between attempts — proven by contention: with a single slot, a second statement submitted mid-backoff stays queued and never reaches the driver
- a full wait queue sheds rather than growing
- the span opens **before** the wait for a slot, so queue time is inside it

The slot-retention test was rewritten after review caught that its first version proved nothing. Sampling `inFlight` from inside the driver reads 1 on every attempt whether the limiter sits outside retry or inside it — the reacquiring arrangement looks identical — so the assertion could not fail when the ordering was wrong, and its comment claimed otherwise. The current version is verified by mutation: moving the limiter inside retry fails it, and passed the original unchanged.

`parseRouteKey.unit.test.ts` covers the six routing cases above.

## Not in this PR

The two follow-ups Part 3 sets up: repositories onto the `QueryRequest` port (the `ClickHouseQueryClient` path, still the design decision described above), and the app's private-route parsing in `clickhouseClient.ts` onto the package's `createTenantRouter`, which already implements it fail-closed.

## Not covered by a test — resolved by Part 3

The log *emission* was untestable while the logger was a module-level singleton. The sinks are now injected, and `vendorClient.test.ts` asserts the emitted records directly: the cluster name on the retry notice (the only record a recovered failure leaves) and on the failure line, and the cold-scan warn against the debug default.

## Test plan

- `pnpm --filter @langwatch/clickhouse-client typecheck` — exit 0
- `pnpm --filter @langwatch/clickhouse-client test:unit` — 185 pass across 10 files, now under `src/__tests__/` like the sibling packages, and now named in `langwatch-app-ci` so the suite actually runs in CI (the app's vitest root cannot reach `packages/`)
- `pnpm test:unit run src/server/clickhouse/ src/server/app-layer/clients/` — 284 pass across 25 suites, the three moved suites among them
- `pnpm --filter @langwatch/web check:feature-parity` — every moved `@scenario` binding still resolves
- `biome check` clean on every touched directory; the one complexity warning on `parsePrivateEnvVars` is pre-existing on `main` and unchanged by this PR

## Summary by CodeRabbit

* **New Features**
  * Added a unified ClickHouse query client with tenant validation, concurrency limits, retries, cancellation, and tracing.
  * Added organization- and cluster-based routing for private ClickHouse connections.
  * Added resilient query handling with metrics, structured logging, error translation, and streamed error detection.

* **Improvements**
  * Improved queue handling when concurrency limits are reached.
  * Enhanced telemetry with execution statistics and safer error details.
  * Prevented sensitive connection credentials from appearing in duplicate-instance logs.

### CODE REVIEW

File: packages/clickhouse-client/src/__tests__/client.test.ts [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+            // deliberately empty of host-specific strings.
+            transientMessageFragments: ["socket hang up"],
+          }),
+        });
+
+        await expect(client.query(request())).resolves.toEqual({ rows: [3] });
+
+        expect(attempts).toBe(3);
+        expect(inFlightPerAttempt).toEqual([1, 1, 1]);
+        expect(limiter.stats()).toEqual({ inFlight: 0, queued: 0 });
```
[the remark this answers is not in the document]
Comment (0xdeafcafe) [reply packages/clickhouse-client/src/__tests__/client.test.ts:134]: Correct, and this was the sharpest finding on the PR. Fixed in 680531f.

Lines 70-72 were false exactly as you describe. With `maxConcurrent: 4` and a single statement in flight, `[1, 1, 1]` is precisely what a limiter placed *inside* the retry loop also produces — it reacquires before each attempt and reads 1 just the same. So the test that supposedly pinned this package's central ordering decision could not fail when that ordering was wrong, while its comment claimed a released slot "would show up as a second acquisition".

Took the shape you suggested: `maxConcurrent: 1`, the retry backoff held open, a second statement submitted while the first is between attempts, asserting it is still queued and never reached the driver.

Verified by mutation rather than assumed — moving the limiter inside retry in `client.ts` fails the new test (`queued` observed 0, expected 1), and passed the old assertion unchanged. Reverted after checking; `client.ts` is byte-identical to what it was.

File: platform/app/src/server/clickhouse/clickhouseClient.ts [hunk 2]
```diff
@@ -58,23 +54,19 @@ function parsePrivateEnvVars(
...
         `Duplicate private ${label} config for orgId "${orgId}": env var "${key}" conflicts with an earlier definition.`,
       );
     }
 
-    map.set(orgId, value);
+    map.set(orgId, { orgId, cluster, url: value });
     logger.info(
-      { orgId, envVar: key },
+      { orgId, cluster, envVar: key },
       `Loaded private ${label} URL from env var`,
```
[the remark this answers is not in the document]
Comment (0xdeafcafe) [reply platform/app/src/server/clickhouse/clickhouseClient.ts:70]: Declining this one — on established platform policy, plus a specific reason redaction would buy nothing at these sites.

**Policy:** LangWatch does not redact identifiers in logs or telemetry. They are opaque surrogate ids, the id→PII mapping is removed on account deletion, and the id is the only thing that makes a log line actionable during an incident. That was settled on #5966 and #5947, including the explicit direction not to add per-site "intentional" comments.

**Specific to these sites:** `orgId` is already emitted on the same log line at both `clickhouseClient.ts#L68-L70` and `#L162-L165`. Hashing or redacting `cluster` there removes the readable name while leaving in place the field that actually identifies the organization — it costs diagnosis and buys no privacy.

**On the label itself:** `cluster` is not one of the fields the path instruction enumerates. It is an operator-authored infrastructure label, chosen by our own infra team inside an env var name, naming a database cluster — and it falls back to the instance name (`shared`) whenever no label was set, which is the common case rather than the exception.

The emission is also the entire point of the change: a rejection from a customer's dedicated instance was indistinguishable from one on the shared cluster, and attributing a three-hour saturation on 2026-08-13 took an afternoon as a result — the answer had to be inferred from a concurrency limit quoted in the vendor's error text and matched against terraform. `resilient-client.ts#L183-L201` already carries a comment saying exactly that.

File: packages/clickhouse-client/src/client.test.ts [hunk 3]
```diff
@@ -0,0 +1,223 @@
...
+       * while the first is between attempts, and must not get it.
+       */
+      it("holds its slot across a retry, so waiting work cannot start between attempts", async () => {
+        const limiter = new ConcurrencyLimiter({ maxConcurrent: 1 });
+        const FIRST = "SELECT 1 FROM t WHERE TenantId = {tenantId:String}";
+        const SECOND = "SELECT 2 FROM t WHERE TenantId = {tenantId:String}";
+
+        const started: string[] = [];
+        let attempts = 0;
+        let sleepEntered = false;
```
[the remark this answers is not in the document]
Comment (0xdeafcafe) [reply packages/clickhouse-client/src/client.test.ts]: Done in 7604008 — renamed to `isRetrySleepEntered` across all three references. 175 package tests still pass.

File: packages/clickhouse-client/src/vendorClient.ts [hunk 4]
```diff
@@ -0,0 +1,562 @@
...
+ * the in-band exception guard for streamed results.
+ *
+ * The vendor package itself is deliberately not imported. Everything here is
+ * declared structurally — a client is anything with `query` and `insert`, a
+ * result is anything with a `json` method — which is what keeps this package
+ * free of the driver dependency and lets a test double be an object literal.
+ * The host supplies what is host-specific through ports: where metrics go,
+ * where log lines go, how a raised error is translated for its callers, and
+ * which queries count as cold scans.
+ */
```
[the remark this answers is not in the document]
Comment (0xdeafcafe) [reply packages/clickhouse-client/src/vendorClient.ts:22]: Done in 45e4a2df94, though the extraction you named was not enough on its own — it came to 346 source lines, because the guards from your other three findings put back roughly what the helpers took out.

Split on two seams instead:

- `statementShape.ts` — the readers you listed (`safeQueryMeta`, `extractQueryType`, `extractTableName`, `extractQueryPreview`, `extractRawQuery`, `inbandExceptionOf`). Pure functions of a vendor value, total against the wrong shape.
- `statementReporting.ts` — the outcome logging, the metric calls and the retry notice, now a `StatementReporter` that the policy class holds.

That second one is the split worth having. One module decides what happens to a statement — retry it, refuse it, guard its stream — and the other decides how it is described. `vendorClient.ts` is 183 source lines and reads as policy.

File: packages/clickhouse-client/src/vendorClient.ts [hunk 5]
```diff
@@ -0,0 +1,562 @@
...
+
+/** Where per-statement outcomes are counted. Omit to count nowhere. */
+export interface StatementMetrics {
+  observeDuration(
+    queryType: VendorQueryType,
+    table: string,
+    durationSeconds: number,
+  ): void;
+  incrementCount(queryType: VendorQueryType, outcome: StatementOutcome): void;
+}
```
[the remark this answers is not in the document]
Comment (0xdeafcafe) [reply packages/clickhouse-client/src/vendorClient.ts]: Done in 45e4a2df94. `StatementMetrics.observeDuration`/`incrementCount` and `translateQueryError` now take an input object, and so does the private `withTransientRetry`. Your reason for doing it now rather than later is the right one: these are exported ports, so the cost of changing them only goes up.

`StatementLogSink` left alone, for the reason you gave.

One consequence worth naming. The platform functions behind those ports — `observeClickHouseQueryDuration`, `incrementClickHouseQueryCount`, `translateClickHouseQueryError` — are older than this PR, positional, and have callers of their own, so they are adapted at `managedClient.ts` rather than rewritten. That function is the seam between what the platform has and what the package asks for; converting them repo-wide is a separate change.

File: packages/clickhouse-client/src/vendorClient.ts [hunk 6]
```diff
@@ -0,0 +1,562 @@
...
+          [QUERY_CAUSE_FIELD]: error,
+        },
+        `ClickHouse ${operation} failed`,
+      );
+    } catch (loggingError) {
+      this.noticeLogger.error(
+        { loggingError },
+        "Failed to log ClickHouse query failure",
+      );
+    }
```
[the remark this answers is not in the document]
Comment (0xdeafcafe) [reply packages/clickhouse-client/src/vendorClient.ts]: Correct, and fixed in 45e4a2df94 — together with the metrics and `onUnscoped` findings, since all three are the same bug in three places.

The fallback is now guarded, and the case you name is what makes it more than belt-and-braces: a host passing one logger for both ports has `noticeLogger === outcomeLogger`, so the sink being asked to report the failure is the sink that just failed. Pinned by a test that passes one broken logger for both and asserts the caller still receives the ClickHouse error, plus one on the success path.

The shape of the fix is not per-call-site `try`, though. Wrapping each call is a discipline, and the next person to add a metric forgets it — so the metrics port and the notice sink are wrapped once on the way into `StatementReporter`. Being unable to throw is now a property of the port rather than of the code using it. The outcome sink stays raw on purpose: its failure is worth a line, and that line goes to the guarded notice sink.

File: packages/clickhouse-client/src/vendorClient.ts [hunk 7]
```diff
@@ -0,0 +1,562 @@
...
+        const durationMs = now() - start;
+        this.logSuccess({ operation: "query", durationMs, params });
+        this.metrics.observeDuration(queryType, table, durationMs / 1000);
+        this.metrics.incrementCount(queryType, "success");
+        return this.guardInbandException({ result, startMs: start, params });
+      } catch (error) {
+        const durationMs = now() - start;
+        this.logFailure({ operation: "query", error, durationMs, params });
+        this.metrics.observeDuration(queryType, table, durationMs / 1000);
+        this.metrics.incrementCount(queryType, "error");
```
[the remark this answers is not in the document]
Comment (0xdeafcafe) [reply packages/clickhouse-client/src/vendorClient.ts]: Agreed, and this was the sharpest of the three. Fixed in 45e4a2df94.

The catch block is exactly the damaging case you describe: `incrementCount` throwing at that line replaced the ClickHouse error, `translateQueryError` never ran, and the caller was handed a metrics failure with no way to learn which ClickHouse error had occurred. The insert paths had it too.

Both ports are now guarded, and guarded separately rather than as a pair — a broken histogram should not suppress a working counter. Four tests pin it: a throwing counter on the read failure path, on the insert failure path, on the success path, and one broken metric beside a working one.

The guard is applied once when `StatementReporter` stores the port, not at each call, so a metric added later inherits it.

File: packages/clickhouse-client/src/tenantGuard.ts [hunk 8]
```diff
@@ -267,17 +267,32 @@ export interface TenantGuardOptions {
...
+   * Returns nothing on success rather than the request: it is a check, and a
+   * caller that had to remember to use a returned value could forget to.
+   */
+  assert(request: QueryRequest): void {
     if (request.unscoped !== undefined) {
-      onUnscoped?.(request);
-      return next(request);
+      this.onUnscoped?.(request);
+      return;
     }
```
[the remark this answers is not in the document]
Comment (0xdeafcafe) [reply packages/clickhouse-client/src/tenantGuard.ts:303]: Right, and fixed in 45e4a2df94. This is the worst-behaved of the three because of which branch it is on: `onUnscoped` runs only where the guard has already decided to allow, so a broken audit sink did not merely lose an audit record — it refused every declared-unscoped statement. A reporting hook deciding policy.

Guarded, with the reason written at the call site pointing at the same rule. Pinned by a test whose `onUnscoped` throws and which asserts the statement still reaches the next layer.

File: packages/clickhouse-client/src/rateLimit.ts [hunk 9]
```diff
@@ -54,48 +54,75 @@ export interface LimiterStats {
...
+  /** Run `task` once a slot is free, releasing the slot when it settles. */
+  async run<T>(task: () => Promise<T>, signal?: AbortSignalLike): Promise<T> {
+    await this.acquire(signal);
+    try {
+      return await task();
+    } finally {
+      this.inFlight -= 1;
+      this.pump();
+    }
   }
```
[the remark this answers is not in the document]
Comment (0xdeafcafe) [reply packages/clickhouse-client/src/rateLimit.ts:108]: Done in 45e4a2df94. `ConcurrencyLimiter.run` now takes `{ task, signal }`, and the two sibling sites you consolidated went with it — the `StatementMetrics` ports and `withTransientRetry` in `vendorClient.ts`. Callers updated: `client.ts`, `statementLimit.ts` and the limiter suite.

File: packages/clickhouse-client/src/tracing.ts [hunk 10]
```diff
@@ -107,63 +107,90 @@ function quietly(report: () => void): void {
...
+    this.spanName = spanName;
+    this.onComplete = onComplete;
+    this.now = now;
+  }
+
+  /** Run `task` inside a span describing `request`. */
+  async trace<Row>(
+    request: QueryRequest,
+    task: () => Promise<QueryResult<Row>>,
+  ): Promise<QueryResult<Row>> {
```
[the remark this answers is not in the document]
Comment (0xdeafcafe) [reply packages/clickhouse-client/src/tracing.ts]: Done in 45e4a2df94 — `QueryTracer.trace({ request, task })`, with `client.ts` and all fourteen call sites in the tracing suite updated.

File: packages/clickhouse-client/src/vendorClient.test.ts [hunk 11]
```diff
@@ -0,0 +1,293 @@
...
+      // The retry notice is the only record a recovered failure leaves, so it
+      // must already name the cluster that refused the attempt.
+      expect(notices.lines).toEqual([
+        expect.objectContaining({
+          level: "warn",
+          fields: expect.objectContaining({ cluster: "acme", attempt: 1 }),
+        }),
+      ]);
+    });
+  });
```
[the remark this answers is not in the document]
Comment (0xdeafcafe) [reply packages/clickhouse-client/src/vendorClient.test.ts]: Fair, and added in 45e4a2df94. You are right that the existing test could not see it: one failure followed by a success passes whether the conversion is `maxRetries + 1`, `maxRetries`, or `maxRetries + 2`, because none of those bounds is reached.

The new test never lets the transient condition clear and asserts the driver was called three times for `maxRetries: 2`. Mutation-checked — changing the conversion to `maxAttempts: this.maxRetries` fails exactly this test and nothing else.
